cip/1.accepted/CIP2016-12-14-Constraint-syntax.adoc

```diff
@@ -0,0 +1,310 @@
+----
+
+==== Compositionality
+
+It is possible to define multiple `REQUIRE` clauses within the scope of the same constraint.
+The semantics between these is that of a conjunction between the constraint expressions of the clauses, such that the constraint is upheld if and only if for all `REQUIRE` clauses, the expression evaluates to `true`.
+
+This is useful not only for readability and logical separation of different aspects of the same constraint, but also for combining the use of the `UNIQUE` operator with other constraint expressions.
+
+=== Examples
+
+In this section we provide several examples of constraints that are possible to express in the specified syntax.
+
+[NOTE]
+The specification in this CIP is limited to the general syntax of constraints, and the following are simply examples of possible uses of the language defined by that syntax. None of the examples provided are to be viewed as mandatory for any Cypher implementation.
+
+Consider the graph created by the statement below.
+The graph contains nodes labeled with `:Color`.
+Each color is represented as an integer-type RGB value in a property `rgb`.
+Users may look up nodes labeled with `:Color` to extract their RGB values for application processing.
+Users may also add new `:Color`-labeled nodes to the graph.
+
+[source, cypher]
+----
+CREATE (:Color {name: 'white', rgb: 0xffffff})
+CREATE (:Color {name: 'black', rgb: 0x000000})
+CREATE (:Color {name: 'very, very dark grey', rgb: 0x000000}) // rounding error!
+----
+
+Owing to the duplication of the `rgb` property, the following attempt at creating a constraint will fail:
+
+[source, cypher]
+----
+CREATE CONSTRAINT only_one_color_per_rgb
+FOR (c:Color)
+REQUIRE UNIQUE c.rgb
+----
+
+Suppose that we would rather like to have one color node per `name` _and_ `rgb` value (to work around the rounding errors).
+We could then use the following constraint, without modifying our data:
+
+[source, cypher]
+----
+CREATE CONSTRAINT unique_color_nodes
+FOR (c:Color)
+REQUIRE UNIQUE c.rgb, c.name
+----
+
+Now, consider the following query which retrieves the RGB value of a color with a given `name`:
+
+[source, cypher]
+----
+MATCH (c:Color {name: $name})
+WHERE exists(c.rgb)
+RETURN c.rgb
+----
+
+The `WHERE` clause is here used to prevent an application from retrieving `null` values for user-defined colors where the RGB values have not been specified correctly.
+It may, however, be eliminated by the introduction of a constraint asserting the existence of that property:
+
+[source, cypher]
+----
+CREATE CONSTRAINT colors_must_have_rgb
+FOR (c:Color)
+REQUIRE exists(c.rgb)
+----
+
+Any updating statement that would create a `:Color` node without specifying an `rgb` property for it would now fail.
+
+Alternatively, we could extend our previous constraint definition with this new requirement:
+
+[source, cypher]
+----
+CREATE CONSTRAINT color_schema
+FOR (c:Color)
+REQUIRE UNIQUE c.rgb, c.name
+REQUIRE exists(c.rgb)
```

Comment:
I'm missing an example for an existence constraint that requires multiple properties to exist on a node; this can be expressed using `exists(n.foo) AND exists(n.bar)`.